File: examples/java-calculator-testng/src/test/java/io/cucumber/examples/testng/ShoppingStepDefinitions.java
package io.cucumber.examples.testng;

import io.cucumber.java.DataTableType;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;

import java.util.List;
import java.util.Map;

import static org.testng.Assert.assertEquals;

public class ShoppingStepDefinitions {

    private final RpnCalculator calc = new RpnCalculator();

    @Given("the following groceries:")
    public void the_following_groceries(List<Grocery> groceries) {
        for (Grocery grocery : groceries) {
            calc.push(grocery.price.value);
            calc.push("+");
        }
    }

    @When("I pay {}")
    public void i_pay(int amount) {
        calc.push(amount);
        calc.push("-");
    }

    @Then("my change should be {}")
    public void my_change_should_be_(int change) {
        assertEquals(-calc.value().intValue(), change);
    }

    @DataTableType
    public Grocery grocery(Map<String, String> entry) {
        return new Grocery(
            entry.get("name"),
            Price.fromString(entry.get("price")));
    }

    static class Grocery {

        private final String name;
        private final Price price;

        Grocery(String name, Price price) {
            this.name = name;
            this.price = price;
        }

    }

    static final class Price {

        private final int value;

        Price(int value) {
            this.value = value;
        }

        static Price fromString(String value) {
            return new Price(Integer.parseInt(value));
        }

    }

}

File: examples/calculator-kotlin-junit5/src/test/kotlin/io/cucumber/examples/calculator/ShoppingStepDefinitions.kt
package io.cucumber.examples.calculator

import io.cucumber.datatable.DataTable
import io.cucumber.examples.calaculator.RpnCalculator
import io.cucumber.java.DataTableType
import io.cucumber.java.en.Given
import io.cucumber.java.en.Then
import io.cucumber.java.en.When
import kotlin.test.assertEquals

class ShoppingStepDefinitions {
    private val rpnCalculator = RpnCalculator()
    private val groceryList = mutableListOf<Grocery>()



    @Given("the following groceries")
    fun givenTheFollowingGroceries(grocery: DataTable){
        val groceries: List<Grocery> = grocery.asList(Grocery::class.java)
        groceryList.addAll(groceries)
        for (gro in groceries){
            rpnCalculator.push(gro.price.value)
            rpnCalculator.push("+")
        }
    }

    @When("I pay {int}")
    fun whenIPay(amount: Int) {
        rpnCalculator.push(amount)
        rpnCalculator.push("-")
    }

    @Then("my change should be {int}")
    fun myChangeShouldBe(change: Int) {
        rpnCalculator.peek()?.let { assertEquals(change, -it.toInt()) }
    }

    @DataTableType
    fun groceryEntry(entry: Map<String, String>): Grocery {
        val name = entry["name"] ?: error("Missing name")
        val price = Price.fromString(entry["price"] ?: error("Missing price"))
        return Grocery(name, price)
    }

    data class Grocery(val name: String, val price: Price)

    @JvmInline
    value class Price(val value: Int) {
        companion object {
            fun fromString(value: String): Price {
                return Price(value.toInt())
            }
        }
    }

}
